src/Client/Client.fs

```diff
@@ -0,0 +1,119 @@
+namespace TcpEcho
+
+open System
+open System.Buffers
+open System.Net.Sockets
+open System.IO.Pipelines
+open System.Text
+open System.Threading.Tasks
+
+type ConnectionUnsuccessfulException(msg: string, innerException: Exception) =
+    inherit Exception(msg, innerException)
+
+exception IncompleteResponseException of string
+type TimeoutOrResult<'a> = Timeout | Result of 'a
+
+[<AbstractClass>]
+type Client(endpoint: string, port: int) =
+    let minimumBufferSize = 1024
+
+    let mkString (buffer: ReadOnlySequence<byte>) =
+        let sb = StringBuilder()
+        for segment in buffer do
+            segment.ToArray() |> Encoding.UTF8.GetString |> sb.Append |> ignore
+        sb.ToString()
+
+    let withTimeout (timeout : int) (xAsync: Async<'a>) = async {
+        let read = async {
+            let! value = xAsync
+            return value |> Result |> Some
+        }
+        let delay = async {
+            do! Async.Sleep(timeout)
+            return Some Timeout
+        }
+
+        match! Async.Choice([read; delay]) with
+        | Some x -> return x
+        | None -> return Timeout
+    }
+
+    let rec writeToPipeAsync (writer: PipeWriter) (socket: Socket) = async {
+        let segment = Array.zeroCreate<byte> minimumBufferSize |> ArraySegment
+        let! read = socket.ReceiveAsync(segment, SocketFlags.None) |> Async.AwaitTask |> withTimeout socket.ReceiveTimeout
+
+        match read with
+        | Timeout ->
+            return writer.Complete(TimeoutException("Socket read timed out"))
+        | Result 0 ->
+            return writer.Complete()
+        | Result bytesRead ->
+            segment.Array.CopyTo(writer.GetMemory(bytesRead))
+            writer.Advance bytesRead
+            let! flusher = writer.FlushAsync().AsTask() |> Async.AwaitTask
+            if flusher.IsCompleted then
+                return writer.Complete()
+            else
+                return! writeToPipeAsync writer socket
+    }
+
+    let rec readFromPipeAsync (reader: PipeReader) str = async {
+        try
+            let! result = reader.ReadAsync().AsTask() |> Async.AwaitTask
+
+            let buffer: ReadOnlySequence<byte> = result.Buffer
+
+            let totalString = str + mkString buffer
+            reader.AdvanceTo(buffer.End)
+
+            match result.IsCompleted with
+            | true ->
+                reader.Complete()
+                return totalString
+            | false ->
+                return! readFromPipeAsync reader totalString
+        with
+            // If we got an TimeoutException anywhere in the exception chain
+            | :? AggregateException as ae when ae.Flatten().InnerExceptions |> Seq.exists (fun x -> x :? TimeoutException) ->
```

Comment:
now that we're not using Task.WhenAny or Task.Delay, can you double check if exceptions still come wrapped under AggregateException?

Reply:
Since it is ```reader.ReadAsync()``` that throws the exception it will be wrapped in an AggregateException.
1) The last rethrow case is not be needed anymore however (aswell as the rethrow in StratumClient#Call).
2) Should the readFromPipeAsync also use the StringBuilder instead of concating strings?

Reply:
1. ok then let's remove them
2. yes please

but, hey, and sorry for the delay, but I rebased your PR somewhere else, because I wanted a CI pipeline to run it, and unfortunately it hangs, so we should fix that first: https://gitlab.com/nblockchain/TcpEchoSharp/merge_requests/1 , let me know your gitlab username and I'll give you access to the repo that holds the PR